src/watcher.zig

```diff
@@ -685,72 +762,133 @@ pub fn NewWatcher(comptime ContextType: type) type {
                     }
                 }
             } else if (Environment.isWindows) {
-                @compileError("watchLoop should not be used on Windows");
-            }
-        }
-
-        pub fn indexOf(this: *Watcher, hash: HashType) ?u32 {
-            for (this.watchlist.items(.hash), 0..) |other, i| {
-                if (hash == other) {
-                    return @as(u32, @truncate(i));
+                var buf: bun.PathBuffer = undefined;
+                const root = this.fs.top_level_dir;
+                @memcpy(buf[0..root.len], root);
+                const needs_slash = root.len == 0 or !bun.strings.charIsAnySlash(root[root.len - 1]);
+                if (needs_slash) {
+                    buf[root.len] = '\\';
                 }
-            }
-            return null;
-        }
+                const baseidx = if (needs_slash) root.len + 1 else root.len;
+                restart: while (true) {
+                    var event_id: usize = 0;
+
+                    // first wait has infinite timeout - we're waiting for the next event and don't want to spin
+                    var timeout = WindowsWatcher.Timeout.infinite;
+                    while (true) {
+                        log("waiting with timeout: {s}\n", .{@tagName(timeout)});
+                        var iter = try this.platform.next(timeout) orelse break;
+                        // after the first wait, we want to start coalescing events, so we wait for a minimal amount of time
+                        timeout = WindowsWatcher.Timeout.none;
+                        const item_paths = this.watchlist.items(.file_path);
+                        log("number of items: {d}\n", .{item_paths.len});
+                        while (iter.next()) |event| {
+                            log("raw event (filename: {}, action: {s})", .{ std.unicode.fmtUtf16le(event.filename), @tagName(event.action) });
+                            var idx = baseidx;
+                            idx += bun.simdutf.convert.utf16.to.utf8.le(event.filename, buf[idx..]);
```

Comment:
This needs to use one of our UTF-16 -> WTF-8  functions because they might have invalid UTF-8 in the filepath